core/src/test/java/kafka/admin/AdminFenceProducersIntegrationTest.java

```diff
@@ -0,0 +1,158 @@
+        config.putIfAbsent(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class.getName());
+
+        return new KafkaProducer<>(config);
+    }
+
+    @ClusterTest
+    void testFenceAfterProducerCommit() throws Exception {
+        clusterInstance.createTopic(TOPIC_NAME, 1, (short) 1);
+
+        producer = createProducer();
+        producer.initTransactions();
+        producer.beginTransaction();
+        producer.send(RECORD).get();
+        producer.commitTransaction();
+
+        adminClient = clusterInstance.createAdminClient();
+        adminClient.fenceProducers(Collections.singletonList(TXN_ID)).all().get();
+
+        producer.beginTransaction();
+        try {
+            producer.send(RECORD).get();
+            fail("expected ProducerFencedException");
+        } catch (ProducerFencedException e) {
+            // expected
+        } catch (ExecutionException e) {
+            assertInstanceOf(ProducerFencedException.class, e.getCause());
+        }
+
+        assertThrows(ProducerFencedException.class, producer::commitTransaction);
+
+        adminClient.close();
+        producer.close();
+    }
+
+    @ClusterTest
+    @Timeout(30)
+    void testFenceProducerTimeoutMs() {
+        Properties config = new Properties();
+        config.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:" + INCORRECT_BROKER_PORT);
+        adminClient = clusterInstance.createAdminClient(config);
+        try {
+            ExecutionException e = assertThrows(ExecutionException.class, () ->
+                    adminClient.fenceProducers(Collections.singletonList(TXN_ID), new FenceProducersOptions().timeoutMs(0)).all().get());
+            assertInstanceOf(TimeoutException.class, e.getCause());
+        } finally {
+            adminClient.close(Duration.ofSeconds(0));
+        }
+
+    }
+
+    @ClusterTest
+    void testFenceBeforeProducerCommit() throws Exception {
+        clusterInstance.createTopic(TOPIC_NAME, 1, (short) 1);
+        producer = createProducer();
+        producer.initTransactions();
+        producer.beginTransaction();
+        producer.send(RECORD).get();
+
+        adminClient = clusterInstance.createAdminClient();
+        adminClient.fenceProducers(Collections.singletonList(TXN_ID)).all().get();
+
+        try {
+            producer.send(RECORD).get();
+            fail("expected Exception");
+        } catch (ExecutionException e) {
+            assertTrue(e.getCause() instanceof ProducerFencedException ||
+                            e.getCause() instanceof InvalidProducerEpochException,
+                    "Unexpected ExecutionException cause " + e.getCause());
+        }
+
+        try {
+            producer.commitTransaction();
+            fail("expected Exception");
+        } catch (ProducerFencedException | InvalidProducerEpochException e) {
+            // expected
+        }
```

Comment:
how about,
```java
ApiException apiException = 
                assertThrows(ApiException.class, producer::commitTransaction, "expected Exception");
        assertTrue(apiException instanceof ProducerFencedException ||
                apiException instanceof InvalidProducerEpochException,
                "Unexpected ApiException " + apiException);
```

Reply:
May I know why use `ApiException` instead of `ExecutionException`